Handle state field requirements for Express Checkout in countries without states

mgascam opened the pull request:
Fixes WOOPMNT-4937
Fixes #10728

This PR addresses an issue with Express Checkout (Apple Pay/Google Pay) where certain countries that don't have state/region fields were causing validation errors in WooCommerce.

#### Changes proposed in this Pull Request

1. Add a new constant COUNTRIES_WITHOUT_STATES in Express_Checkout_Element_States class to identify countries where WooCommerce provides a state field but the Express Checkout Element API does not
2. Implement a new filter woocommerce_get_country_locale to modify country locale settings specifically for Express Checkout contexts
3. Make the state field optional and hidden for affected countries during Express Checkout to prevent validation errors

This solution was chosen because:

- It's a non-intrusive way to handle the state field requirements
- It only affects Express Checkout flows, preserving the normal checkout behavior

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Pre-requisites**
1. Enable the `Enable Cart-Token implementation for ECE ` checbox in WCPay Dev tool
2. Enable Apple Pay / Google Pay in Payments > Settings
3. `npm run tube:start` so the Express Checkout buttons are available

**Testing with Google Pay**
1. With Google Chrome, Go to a page where the Express Checkout is available: product, cart or checkout.
2. Click the Google Pay button
4. In the Google Pay modal, select and address from an affected country i.e. Romania

<img width="610" alt="Screenshot 2025-05-09 at 16 04 10" src="https://github.com/user-attachments/assets/c0e829e1-44da-4349-805c-b20a52943180" />

5. Click the Pay now button to proceed with the payment.
6. Check that the order is correctly processed and there are no state/county related errors.

**Testing with Apple Pay**
On top of the other requirements you'll need:

- Use Safari browser for testing.
- You'll need an Apple ID from an affected country or you can use an Apple Sandbox account. There are instructions available here pcrvl0-1yh-p2.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Add express checkout country locale modifications and new constants for countries without states

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10770` or branch name `woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: e87d9a4f6aa79c224c813507503dfd62c1807e5f
- Build time: 2025-05-21 11:06:42 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Add changelog

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

htdat commented on includes/constants/class-express-checkout-element-states.php:
```diff
@@ -1159,4 +1159,35 @@ class Express_Checkout_Element_States {
+	/**
+	 * List of countries where WooCommerce provides a state/region field but the Express Checkout
+	 * Element API does not. For these countries, we need to make the
+	 * state field optional and hidden to avoid validation errors.
```
What's the source for Express Checkout Element API? I would like to have a look at them. 

If Apple Pay / Google Pay do not require `state/county` for a specific country, Woo core should also revisit itself?

frosso replied:
If possible, it would be great to document where this list of countries has been retrieved :) did you manually do through the countries in Apple Pay, and looked at each that didn't provide a state? or were you able to get some documentation about this?

mgascam replied:
Thanks for double-checking the list of countries. I tried to find some official documentation from Apple or Google but couldn't find anything. Fortunately, I found a reference to libaddressinput in the [codebase](https://github.com/Automattic/woocommerce-payments/blob/737886b4517db5dec4a3f1cb9fe4f9fb89460880/includes/constants/class-express-checkout-element-states.php#L16), along with a list of countries where the address form doesn’t include a state field. I used that list as a starting point and manually confirmed that the native address form in the Mac wallet omits the state field for those countries.

mgascam replied:
> What's the source for Express Checkout Element API? I would like to have a look at them.

That comment was a bit misleading — it’s not the Express Checkout Element API, but rather the native address form used to enter address details that lacks the state field for some countries. I’ve updated the comment to clarify this.

> If Apple Pay / Google Pay do not require state/county for a specific country, Woo core should also revisit itself?

Great point, that was my initial thought as well. I tested making the state field optional in Woo core, and it did resolve the issue. However, I was hesitant to make a core change for something that feels quite specific to WooPayments and Apple Pay / Google Pay. So instead, I opted to use the `woocommerce_get_country_locale` hook to adjust field requirements only during Apple/Google Pay checkout.

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-button-helper.php:
```diff
@@ -1250,4 +1249,109 @@ protected function calculate_shipping( $address = [] ) {
+	 *
+	 * @return bool True if this is an express checkout API request, false otherwise.
+	 */
+	private function is_store_api_request() {
```
We have already had a few utils for this purpose, I don't think we need to have it here such as: 

- WC_Payments_Utils::is_store_api_request
- WCPay\MultiCurrency\Utils::is_store_api_request - from Multicurrency module, we should not use it as it's not always available but at least we can use `WC_Payments_Utils::is_store_api_request`. 
- WC()->is_store_api_request() - from Woo core.

frosso replied:
I agree - I think that `\WC_Payments_Utils::is_store_api_request()` could be used instead, to be consistent with the other areas of the codebase?

I think we have `\WC_Payments_Utils::is_store_api_request()` because `WC()->is_store_api_request()` in WC Core is relatively new.

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-button-helper.php:
```diff
@@ -1250,4 +1249,109 @@ protected function calculate_shipping( $address = [] ) {
+
+			if ( 'payment_request_type' === $payment_data_item['key'] || 'express_payment_type' === $payment_data_item['key'] ) {
+				return in_array( $payment_data_item['value'], [ 'apple_pay', 'google_pay' ], true );
+			}
```
The logic within this foreach loop can be simplified like the following: 

```php
			$key = $payment_data_item['key'] ?? null;
			$value = $payment_data_item['value'] ?? null;

			if (
				in_array( $key, ['payment_request_type', 'express_payment_type'], true )
				&& in_array( $value, [ 'apple_pay', 'google_pay' ], true )
			) {
				return true;
			}
```

htdat commented on includes/express-checkout/class-wc-payments-express-checkout-button-helper.php:
```diff
@@ -1250,4 +1249,109 @@ protected function calculate_shipping( $address = [] ) {
+		$overrides = $this->get_locale_overrides();
+		foreach ( $overrides as $country_code => $override ) {
+			$locales[ $country_code ] = $override;
+		}
```
IMO, we would only change the most relevant arg `required` of `$locales`, so we would try to be as much granular as possible. 

```suggestion
		// No longer needs the `get_locale_overrides` method. 
		foreach ( Express_Checkout_Element_States::COUNTRIES_WITHOUT_STATES as $country_code ) {
			$locales[ $country_code ]['state']['required'] = false;
		}
```

htdat reviewed (commented):
I am only manually testing with Google Pay but I can not replicate the problem mentioned in the Linear issue. 

Instead, I only get this error - [screenshot](https://github.com/user-attachments/assets/5f8f872a-67f8-4b54-97a9-d776eb0d7e32): 
![Markup on 2025-05-15 at 14:01:49](https://github.com/user-attachments/assets/2db271f3-699f-4bfa-bef7-9245cfe1bd7f)



> The express checkout is not supported in United States (US) because some required fields couldn't be verified. Please proceed to the checkout page and try again. Shipping County is a required field.

When applied this PR, I am still getting the same error. 

Note that: I chose both my Woo customer account's billing country and the address on Google Pay model are Romania.

frosso commented on includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:
```diff
@@ -95,6 +95,7 @@ public function init() {
 		add_action( 'before_woocommerce_pay_form', [ $this, 'display_pay_for_order_page_html' ], 1 );
 		add_filter( 'woocommerce_gateway_title', [ $this, 'filter_gateway_title' ], 10, 2 );
 		add_action( 'woocommerce_checkout_order_processed', [ $this->express_checkout_helper, 'add_order_payment_method_title' ], 10, 2 );
+		add_filter( 'woocommerce_get_country_locale', [ $this->express_checkout_helper, 'modify_country_locale' ], 20 );
```
There is a `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`, which is used to handle most (if not all?) of the incoming AJAX requests for the express checkout buttons. I think `WC_Payments_Express_Checkout_Ajax_Handler` would be a good spot for this filter, since it's used on AJAX requests.

What do you think about using that class for this filter?

frosso reviewed (commented):
Great work with using that filter for the Store API request when coming from the ECE.
I tested the changes locally, but I encountered some issues.

On `develop`, this is the error I get with a romanian address using Apple Pay:
<img width="1137" alt="Screenshot 2025-05-16 at 4 41 47 PM" src="https://github.com/user-attachments/assets/3364735a-84ab-41f1-b570-76ac487ff781" />

On this branch, instead, this is the error I get:
<img width="1145" alt="Screenshot 2025-05-16 at 4 38 12 PM" src="https://github.com/user-attachments/assets/042bc41c-87b0-4b29-8d5d-e19a6a8903da" />

I got this address from [the random address generator](https://www.fakeaddressgenerator.com/All_countries/address/country/Romania):
<img width="825" alt="Screenshot 2025-05-16 at 4 46 06 PM" src="https://github.com/user-attachments/assets/d2b7ac02-ddbd-4413-aed5-bcc0712fda23" />

And I entered it in Apple Pay:
<img width="706" alt="Screenshot 2025-05-16 at 4 45 49 PM" src="https://github.com/user-attachments/assets/37c8ff49-e212-4718-8827-f991a08f3aa2" />
<img width="376" alt="Screenshot 2025-05-16 at 4 45 01 PM" src="https://github.com/user-attachments/assets/d1a6731d-8ed4-40b1-b730-becfc9efe0eb" />
<img width="390" alt="Screenshot 2025-05-16 at 4 44 40 PM" src="https://github.com/user-attachments/assets/b3e75d9e-8d11-4d42-a303-26d73663d63d" />


I also tried with a PK address, as outlined in Slack, and I get a similar error:
<img width="1120" alt="Screenshot 2025-05-16 at 4 50 20 PM" src="https://github.com/user-attachments/assets/d6c1922d-bbd5-487e-af09-85befa579711" />
<img width="390" alt="Screenshot 2025-05-16 at 4 51 29 PM" src="https://github.com/user-attachments/assets/7ccc37cc-b413-4db9-83c9-df0cb8ee05df" />

So maybe we need to skip validation also for the zip code?

mgascam committed:
PR feedback: replace local store API request check with utility function in express checkout button helper

mgascam committed:
PR feedback: simplify foreach loop logic

mgascam committed:
PR feedback: streamline locale handling for countries without state fields

mgascam committed:
PR feedback: place country locale filter in the WC_Payments_Express_Checkout_Ajax_Handler

mgascam commented:
Thanks @frosso and @htdat for your review and testing, much appreciated!. I addressed most of your comments already, but still need to look into the issues you found. Will let you know when I'm ready for a new review round.

mgascam commented:
> I am only manually testing with Google Pay but I can not replicate the problem mentioned in the Linear issue.
> 
> Instead, I only get this error - [screenshot](https://github.com/user-attachments/assets/5f8f872a-67f8-4b54-97a9-d776eb0d7e32):

Thanks for testing, @htdat! I was able to reproduce the same issue you described by disabling the `Enable Cart-Token implementation for ECE` feature flag in the WCPay Dev tool:

<img width="801" alt="Screenshot 2025-05-19 at 11 54 27" src="https://github.com/user-attachments/assets/fb6be3d6-82e3-48b1-be46-e5b7a44c5a3c" />

Looking at this [method](https://github.com/Automattic/woocommerce-payments/blob/1e426d03c169582c100c23780f85237da85db366/includes/class-wc-payments-features.php#L51) — and also this [PR](https://github.com/Automattic/woocommerce-payments/pull/10415) — this feature is enabled by default unless explicitly set to false. The WCPay Dev tool disables it by default, which doesn't match the current behavior in production.

Could you please try enabling the Enable Cart-Token implementation for ECE flag in the dev tool and check again? I’ve also updated the PR description to clarify this.

Additionally, I’ll look into enabling this flag by default in the WCPay Dev tool so it aligns with production behavior.

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

mgascam commented:
> So maybe we need to skip validation also for the zip code?

Thanks for testing, @frosso! In my opinion, this looks like an inconsistent or buggy implementation in the Apple Wallet address form. I would expect the form validation to fail due to the missing postal code, since the field is present in the form and not marked as optional.

I tried your suggestion of using the filter to also mark the postal code as optional, and it does work — the payment goes through and the order is placed. However, I’m hesitant to adopt this change because the postal code seems important for merchants, especially for shipping purposes.

I see this as different from skipping the state validation for two main reasons:

The postal code is more critical, as the state can often be inferred from it.

Customers can manually add a postal code to their Apple Wallet address, whereas the state field isn’t even present, so there’s no way for customers to fix the issue themselves.

What do you think?

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

frosso commented on includes/class-wc-payments.php:
```diff
@@ -488,6 +488,7 @@ public static function init() {
 		include_once __DIR__ . '/constants/class-payment-capture-type.php';
 		include_once __DIR__ . '/constants/class-payment-method.php';
 		include_once __DIR__ . '/constants/class-track-events.php';
+		include_once __DIR__ . '/constants/class-express-checkout-element-states.php';
```
Why move this files from a conditional `include_once` into being always included?
I think one of the benefits from conditionally including it was to avoid it being loaded when not needed, since it is a large key-value-pair that would otherwise need to be loaded in memory for each request.
I was looking at the query monitor, and noticed an increase of ~500KB of memory usage, compared to `develop`.

mgascam replied:
Great catch, thanks for pointing it out since I was not aware of the drawback. I'm going back to conditionally including it in this commit [bf84367](https://github.com/Automattic/woocommerce-payments/pull/10770/commits/bf84367a84472a4286ee6779306fe2ee051d4161)

frosso commented on includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:
```diff
@@ -95,6 +95,7 @@ public function init() {
 		add_action( 'before_woocommerce_pay_form', [ $this, 'display_pay_for_order_page_html' ], 1 );
 		add_filter( 'woocommerce_gateway_title', [ $this, 'filter_gateway_title' ], 10, 2 );
 		add_action( 'woocommerce_checkout_order_processed', [ $this->express_checkout_helper, 'add_order_payment_method_title' ], 10, 2 );
+		add_filter( 'woocommerce_get_country_locale', [ $this->express_checkout_ajax_handler, 'modify_country_locale_for_express_checkout' ], 20 );
```
What do you think about moving this filter registration to `includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php`, in its `init` method? This way the `modify_country_locale_for_express_checkout` method` can stay colocated with the filter registration.

mgascam replied:
I agree, the thing is I tried once and the filter was not being triggered. But probably I was doing something wrong. I tried again and it worked fine so I refactored it in this commit [7937c7c](https://github.com/Automattic/woocommerce-payments/pull/10770/commits/7937c7c76f49505a74785ce34526431ac8220e8c)

frosso commented on includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:
```diff
@@ -606,4 +606,76 @@ private function transform_ece_address_postcode_data( $address ) {
+		$request_body = file_get_contents( 'php://input' );
+		$request_data = json_decode( $request_body, true );
+
+		if ( ! isset( $request_data['payment_data'] ) ) {
+			return false;
+		}
+
+		return $this->is_express_payment_type( $request_data['payment_data'] );
```
(minor and optional :) ) one alternative to these checks could be to implement similar checks as the ones defined on top of the `tokenized_cart_store_api_address_normalization` method: https://github.com/Automattic/woocommerce-payments/blob/c8bad250e705bfa62c68037de94bd3332ca25ce9/includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php#L500-L508

This way, you don't need to inspect the request's payment data, and you can use the nonce validation that is already used in other methods.

mgascam replied:
I think that's a better way thanks for the suggestion. I tried the same approach in this commit [96d5616](https://github.com/Automattic/woocommerce-payments/pull/10770/commits/96d561697567a38fa649d9e111c6c629f36a05a0). The only difference is that   we don't have direct access to the request in the `woocommerce_get_country_locale` filter, so I'm using superglobals to access it. Please have a look and let me know what you think.

frosso replied:
Using the superglobals seems fine, since `$request` isn't filled yet. We need to use those same superglobals in `WC_Payments_Payment_Request_Session` for the same reason :)

frosso reviewed (approved):
A few suggestions for consideration, but nothing major.
I tested this again, and it also works with the Romanan user that is part of the Google Pay test addresses.
I like the workaround you implemented that uses the `woocommerce_get_country_locale` filter to mark the ZIP code as "optional" for some countries when the ECE buttons are used as payment methods.
You also convinced me and made a good argument about the scenario where the ZIP code is missing.

In https://github.com/Automattic/woocommerce-payments/pull/10784 I implemented the removal of the Tokenized ECE feature flag, so that it shouldn't create confusion anymore :)

mgascam committed:
Include states constants only when needed

mgascam committed:
Refactor is_express_checkout_context to use tokenized cart header and nonce

mgascam committed:
Refactor initialization of the woocommerce_get_country_locale filter

mgascam committed:
Amend comment

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing

mgascam commented:
@frosso @htdat I believe I addressed all your latest comments, ready for a new review round when you have a moment 🙏

frosso reviewed (approved):
Tested again, still looking great! I can confirm that on `develop` I get the error message when attempting to checkout, but I don't get the message on this branch and I can checkout successfully!
<img width="779" alt="Screenshot 2025-05-21 at 12 39 02 PM" src="https://github.com/user-attachments/assets/8408c4cf-b39f-43d6-a2e6-22aec97bb963" />

Thank you again for the changes and for your patience with this! :shipit:

mgascam committed:
Merge branch 'develop' into woopmnt-4937-apple-pay-checkout-fails-with-woopayments-when-billing
